subsys/usb/usb_descriptor.c

```diff
@@ -290,16 +291,30 @@ static struct usb_cfg_data *usb_get_cfg_data(struct usb_if_descriptor *iface)
 }
 
 /*
- * Default USB SN string descriptor is CONFIG_USB_DEVICE_SN, but sometimes
- * we want use another string as SN descriptor such as the chip's unique
- * ID. So user can implement this function return a string thats will be
- * replaced the default SN.
- * Please note that the new SN descriptor you changed must has same length
- * as CONFIG_USB_DEVICE_SN.
+ * Default USB Serial Number string descriptor will be derived from
+ * Hardware Information Driver (HWINFO). User can implement own variant
+ * of this function. Please note that the length of the new Serial Number
+ * descriptor may not exceed the length of the CONFIG_USB_DEVICE_SN.
  */
 __weak u8_t *usb_update_sn_string_descriptor(void)
 {
-	return NULL;
+	u8_t hwid[sizeof(CONFIG_USB_DEVICE_SN) / 2];
+	static u8_t sn[sizeof(string_descr.utf16le_sn.bString)];
+	const char hex[] = "0123456789ABCDEF";
+
+	BUILD_ASSERT((sizeof(hwid) * 4) <= sizeof(sn));
+	memset(hwid, 0, sizeof(hwid));
+	memset(sn, 0, sizeof(sn));
+
+	if (hwinfo_get_device_id(hwid, sizeof(hwid)) > 0) {
```

Comment:
I think @aurel32 is right here. Maybe we should consider using a CRC.

Reply:
Maybe `hwinfo_get_device_id_16`, `hwinfo_get_device_id_32`, `hwinfo_get_device_id_64`?
This could be hardware dependent then. Nordic chips, for example, could just return parts and STM32 do the CRC.

Reply:
or hwinfo_get_device_id_hash() based on length of the buffer? But is crc32 sufficient for that?

Reply:
CRC is sufficient if you have a fixed length.